Add pre and post snapshot and prune hooks

sidevesh opened the pull request:
<!--
Before submitting a PR, please read the Contributing Guidelines:
https://github.com/zhangyuannie/butter/blob/main/CONTRIBUTING.md
-->

<!-- To check a box, change "[ ]" to "[x]" -->

- [x] I have read and agree to the Contributing Guidelines.
- [x] This compiles and is ready for review.

<!--
If your PR is linked to a issue, please reference it below.
e.g. fix #5, close #12
-->
Closes #15

sidevesh committed:
Add pre and post snapshot and prune hooks

sidevesh committed:
Remove unused ffi::OsStr use

zhangyuannie commented:
Thanks for the amazing work! I'm quite busy recently unfortunately, will hopefully take a look soon (ETA >= 2w).

sidevesh commented:
@zhangyuannie were you able to take a look ?

zhangyuannie commented on src/rule.rs:
```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
             let target_path = self.target_dir.join(name.as_str());
+            if !rule.pre_snapshot.is_empty() {
+                let mut cmd = std::process::Command::new("sh");
+                cmd.arg("-c").arg(format!("{} {} {}", rule.pre_snapshot, self.path.display(), target_path.display()));
```
We probably also need to document this somewhere. We probably want to also have an option for letting this script run in root/user mode. For now I think you can do a simple check to make sure this file is only writable by root

sidevesh replied:
Is the check if the script is only writable by root so that someone without root can't use it for executing things as root ?
That would be an issue though even if you meant the check for something else, I will add that.

zhangyuannie replied:
It's so that someone without root can't modify the script.

zhangyuannie commented on src/rule.rs:
```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
+                            } else {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, error: {}", String::from_utf8_lossy(&output.stderr));
+                            }
+                            return Err(anyhow::anyhow!("Pre snapshot script failed"));
```
should we abort snapshotting because user's script failed?

sidevesh replied:
my usecase here is to use the pre snapshot to check / mount the external backup location,
local snapshots are useless for me as a backup thing if it is not kept externally,
using a pre action script to check things before letting action happen is a pattern I have seen and could be useful for other usecases.
With btrfs snapshots being so cheap, there isnt really much difference otherwise between pre and post, except that pre can prevent backup from happening if so desired.

zhangyuannie reviewed (commented):
I started looking at this last week but wasn't able to finish. There's quite a lot of details.

TLDR: I'm ok shipping this if:

- We change the json schema from `pre_snapshot` to `pre_snapshot_EXPERIMENTAL`, etc.
- We at least make sure the script we ran is only writable by root (or run it as user instead)
- Maybe not showing on the UI yet (or just the post snapshot hook to start with)

sidevesh commented on src/rule.rs:
```diff
@@ -232,12 +284,43 @@ impl RuleSubvolume {
             }
 
             if should_remove {
+                if !rule.pre_prune.is_empty() {
```
I agree, pre / post prune is gonna be very useless for most people,
my use case for it is to cleanup snpshots on the external backup once its pruned from the system,
there maybe other cleanup related stuff that user might want to do when a snapshot is pruned but tbh I can't think of any other usecase either.

sidevesh commented on src/rule.rs:
```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
+                Ok(_) => {
+                    if !rule.post_snapshot.is_empty() {
+                        let mut cmd = std::process::Command::new("sh");
+                        cmd.arg("-c").arg(format!("{} {} {}", rule.post_snapshot, self.path.display(), target_path.display()));
```
Post snapshot hook here should be getting the path to the snapshot.
Taking custom command with support for placeholders for the variables we insert could be useful for quick things but I don't think it would make much difference right now.

sidevesh commented on data/resources/ui/schedule_rule_edit_dialog.ui:
```diff
@@ -192,6 +192,74 @@
+
+                    <child>
+                      <object class="AdwActionRow">
+                        <property name="title" translatable="yes">Pre Snapshot</property>
```
I agree, my goal with this PR was to experiment and try out ways to move and manage backups on external location,
and that needed all the 4 script hooks that I added.
I am trying to model my setup similar to Time Machine on macOS and its backup and cleanup scheme for snapshots and external backups.
And I have realized that a backup tool like that does not fit Butter's workflow exactly,
I think making the pre / post prune fields experimental and not showing them in UI right now would make sense right now,
I think we should keep the pre snapshot as it might be useful in some usecases and pre / post hooks around an action is a common pattern in general but also in other backup tools.

sidevesh commented:
@zhangyuannie are you okay with accepting this with following changes:
  -- Only run scripts if they are only writable by root, otherwise fail, add to the message in section subtitle informing about this
  -- Make pre / post prune hooks `_EXPERIMENTAL` in json schema and hide it from UI

zhangyuannie commented:
Please also add _EXPERIMENTAL to pre and post snapshot hooks. They can show up on the UI if its label also mention that this feature is experimental.

This is to avoid commiting to something that we will regret. For example, you mentioned:

> Post snapshot hook here should be getting the path to the snapshot.

However, in your current code, that is not the case.

zhangyuannie commented:
Hi @sidevesh, are you still interested in moving this forward?

sidevesh commented:
hey @zhangyuannie , I am but I wasn't able to clean this up due to losing access to my linux laptop while travelling.
I will look into addressing the points in next weeks sometime once I am back with my laptop.

sidevesh commented:
I have had quite a bit of success with making Butter move snapshots to a sparsefile kept in samba share using these pre post snapshot hooks,
I was working on moving that process out from the script that I assign to the hooks to Butter's rust code itself,
allowing the user to specify the network share details and location to keep the sparsefile disk image on and then moving the snapshot after the backup.
So my original need for hooks is now solved in my wip fork of Butter, still makes sense to get atleast pre post snapshot hooks  in as that's anyways a standard thing in most backup solutions.

sidevesh commented:
Question, would you be interested in upstreaming the network backup functionality I am working on in my fork into Butter itself ?

zhangyuannie commented:
> allowing the user to specify the network share details and location to keep the sparsefile disk image on and then moving the snapshot after the backup.

I have very limited experience with sparsefile and samba so would need to understand how your solution works. So ideally I can have a section of "community supported features/extensions" and just dump it there but probably not gonna happen anytime soon if at all. :stuck_out_tongue_closed_eyes:

So I would say let's have pre/post snapshot/prune hooks first with `_EXPERIMENTAL` suffix everywhere. Thanks!
